### PR TITLE
Fix #7: use intel_pstate/no_turbo to disable turboboost

### LINKED ISSUE #7: Use `intel_pstate/no_turbo` to disable turboboost in preference to msr-tools
Currently the uarch-bench.sh wrapper script uses `rdmsr` and `wrmsr` from msr-tools to disable turbo by writing the msr regs, but a cleaner way for systems using the `intel_pstate` CPU governor (probably most modern systems) would be to write to `/sys/devices/system/cpu/intel_pstate/no_turbo` instead. Simpler (no need to do everything per core) and doesn't require the user to install `msr-tools`.

### PR DESCRIPTION
**Pull request by jinankjain**
Using this would resolve the dependency on `msr-tools` to disable turbo boost.

### CODE REVIEW

File: uarch-bench.sh [hunk 1]
```diff
@@ -4,41 +4,15 @@
...
-function get_turbo_state_string {
-	local turbo_state
-	for core in $cores; do
-    	turbo_state+="$(get_core_turbo_intel $core) "
-	done
-	echo $turbo_state
-}
-
-
-function is_even {
```
Comment (travisdowns) [inline uarch-bench.sh]: Oops, thanks for deleting this, this was just spuriously left over from some testing I was doing :).

File: uarch-bench.sh [hunk 2]
```diff
@@ -4,41 +4,15 @@
...
-  if [ $(( $1 % 2)) -eq 0 ]; then echo "O";
-  else echo "X";
-  fi
-}
-
 export VENDOR_ID=$(lscpu | grep 'Vendor ID' | egrep -o '[^ ]*$')
 export MODEL_NAME=$(lscpu | grep 'Model name' | sed -n 's/Model name:\s*\(.*\)$/\1/p')
 export SCALING_DRIVER=$(cat /sys/devices/system/cpu/cpu0/cpufreq/scaling_driver)
 export SCALING_GOVERNOR=$(cat /sys/devices/system/cpu/cpu0/cpufreq/scaling_governor)
+export NO_TURBO=$(cat /sys/devices/system/cpu/intel_pstate/no_turbo)
```
Comment (travisdowns) [inline uarch-bench.sh]: See the review overall comments, but we should handle the case where this path doesn't exist.

**Comment by jinankjain** [comment]
@travisdowns Updated the Pull Request with the suggested changes :smile:

**Comment by travisdowns** [comment]
@jinankjain - thanks! I rebased and merged this in 277ff7725ac37a6c0e7a72e923cf155623c60781. I think this PR won't close automatically due to the rebase (different SHAs), so closing this now. 

Thanks again for your help and I look forward to any further contributions!

**Comment by jinankjain** [comment]
Thanks @travisdowns for merging this. I was thinking of working on porting this MacOSX. I saw that there is an open issue.

**Comment by travisdowns** [comment]
@jinankjain - that would be awesome!

The primary difficulty for "full functionality", I think, would be porting over the `libpfc` stuff which allows you to read the performance counters. If one were to simply use the default `ClockTimer` implementation, then the code is pretty much portable C++ (plus nasm asm which no doubt is available on OSX - the asm stuff should pretty much work "as is" since the sysv calling convention is used there also), and so the initial port should be straightforward. Beyond that there is the matter of disabling turbo and other frequency scaling (the stuff in `uarch-bench.sh` basically), where the details are no doubt different.

**Comment by jinankjain** [comment]
@travisdowns How about using https://github.com/opcm/pcm instead of libpfc as it provides much generic interface for Mac, Windows or Linux etc?

**Comment by travisdowns** [comment]
@jinankjain yes, sure. In fact I recall now that this was mentioned on the [associated issue](https://github.com/travisdowns/uarch-bench/issues/2). I don't think you can "replace" `libpfc` with `pcm` because the latter doesn't offer user-mode `rdpmc` functionality, but you could certainly use the kernel driver parts to enable the counters to avoid re-inventing the wheel. In principle, the user-mode parts of `libpfc` (basically the `PFC_END` macro) should carry over pretty much unchanged in OSX since they are just very small inline assembly wrappers around the `rdpmc` instruction, which should be OS-independent (as long as the bit in `CR4` is set that enables user-mode `rdpmc` access).

**Pull request closed without merging**